### PR TITLE
Add batch sampler to DataLoader

### PR DESCRIPTION
**Pull request by colesbury**
The optional batch_sampler argument combines the sampler with the
grouping of indices into mini-batches. This allows for non-standard
batch sizes.

This is important for some sequence to sequence tasks. For example, when [fairseq](https://github.com/facebookresearch/fairseq/) evaluates on the validation or test set, it groups samples into batches only if it would not require padding. This means that some batches may be smaller than the requested batch size.

### CODE REVIEW

File: torch/utils/data/dataloader.py [hunk 1]
```diff
@@ -272,38 +254,44 @@ class DataLoader(object):
             at every epoch (default: False).
         sampler (Sampler, optional): defines the strategy to draw samples from
             the dataset. If specified, the ``shuffle`` argument is ignored.
+        batch_sampler (Sampler, optional): like sampler, but returns a batch of
```
Comment (apaszke) [inline torch/utils/data/dataloader.py]: It would be nice to raise an error when both batch_sampler and one of the ignored options is specified. Just change their default values to None and add checks

File: torch/utils/data/dataloader.py [hunk 2]
```diff
@@ -272,38 +254,44 @@ class DataLoader(object):
...
+            into CUDA pinned memory before returning them.
         drop_last (bool, optional): set to ``True`` to drop the last incomplete batch,
             if the dataset size is not divisible by the batch size. If False and
             the size of dataset is not divisible by the batch size, then the last batch
             will be smaller. (default: False)
     """
 
-    def __init__(self, dataset, batch_size=1, shuffle=False, sampler=None, num_workers=0,
-                 collate_fn=default_collate, pin_memory=False, drop_last=False):
+    def __init__(self, dataset, batch_size=1, shuffle=False, sampler=None, batch_sampler=None,
```
Comment (apaszke) [inline torch/utils/data/dataloader.py]: Inserting an arg in the middle is a backward incompatible change.

Comment (colesbury) [inline torch/utils/data/dataloader.py]: I think it's unlikely to break anyone in this case and it's nicer to have it next to sampler in the docs.

**Review by apaszke — APPROVED** [review]
